In [123]:
### PART 1 Working with JSON data

In [1]:
# pip install pooch requets

In [2]:
import requests
import json

In [10]:
import json
import requests
# Here's a sample JSON structure similar to what APIs return
sample_json = '''
{
  "station": "USC00305800",
  "name": "New York Central Park",
  "location": {
    "latitude": 40.7789,
    "longitude": -73.9692
  },
  "observations": [
    {"date": "2023-01-01", "temperature": 32, "precipitation": 0.0},
    {"date": "2023-01-02", "temperature": 28, "precipitation": 0.5},
    {"date": "2023-01-03", "temperature": 35, "precipitation": 0.0},
    {"date": "2023-01-04", "temperature": 38, "precipitation": 0.2},
    {"date": "2023-01-05", "temperature": 41, "precipitation": 0.0}
  ]
}
'''

# Parse the JSON
data = json.loads(sample_json)

# Access nested data
print("Station:", data['station'])
print("Location:", data['location'])
print("First observation:", data['observations'][0])

Station: USC00305800
Location: {'latitude': 40.7789, 'longitude': -73.9692}
First observation: {'date': '2023-01-01', 'temperature': 32, 'precipitation': 0.0}


In [12]:
# 1. Extract and print all dates and temperatures (8 points)

In [26]:
print("Date, Temperature")
for obs in data['observations']:
    print(f"{obs['date']}, {obs['temperature']}")

Date, Temperature
2023-01-01, 32
2023-01-02, 28
2023-01-03, 35
2023-01-04, 38
2023-01-05, 41


In [20]:
# 2. Calculate average temperature

In [29]:
total_temp = 0
count = 0
for obs in data['observations']:
    total_temp += obs['temperature']
    count += 1

avg_temp = total_temp / count
print(f"Average temperature: {avg_temp}°F")

Average temperature: 34.8°F


In [30]:
# 3. Find days with precipitation

In [35]:
print("\nDays with precipitation:")
for obs in data['observations']:
        # Check if precipitation is greater than 0
    if obs['precipitation'] > 0:
        print(f"Date: {obs['date']} | Precipitation: {obs['precipitation']} in")


Days with precipitation:
Date: 2023-01-02 | Precipitation: 0.5 in
Date: 2023-01-04 | Precipitation: 0.2 in


In [38]:
# Now try with a real API 
# Use a real weather API (you may need to sign up for a free API key)
# Example APIs: OpenWeatherMap, NOAA, Weather.gov
# Used Weather.gov since it doesn't require API key

In [81]:
# Import data from api.weather.gov using this code from webstite 

In [82]:
import requests
import json
{
  "@context": [
    "string"
  ],
  "type": "FeatureCollection",
  "features": [
    {
      "@context": [
        "string"
      ],
      "id": "string",
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          0,
          0
        ],
        "bbox": [
          0,
          0,
          0,
          0
        ]
      },
      "properties": {
        "@context": [
          "string"
        ],
        "geometry": "string",
        "@id": "string",
        "@type": "wx:ObservationStation",
        "elevation": {
          "value": 0,
          "maxValue": 0,
          "minValue": 0,
          "unitCode": "e0@A@9JrhssKVm:/+uD$8iW0v&+`jT*)Kdh<S@)R4`>Zf\"X.|_1%=85_mxd-b@>Qo ep^fJe\\.uuEA9EhXXd",
          "qualityControl": "Z"
        },
        "stationIdentifier": "string",
        "name": "string",
        "timeZone": "string",
        "provider": "string",
        "subProvider": "string",
        "forecast": "string",
        "county": "string",
        "fireWeatherZone": "string",
        "distance": {
          "value": 0,
          "maxValue": 0,
          "minValue": 0,
          "unitCode": "uc:0k%4=i.gN%!{nFJfyl",
          "qualityControl": "Z"
        },
        "bearing": {
          "value": 0,
          "maxValue": 0,
          "minValue": 0,
          "unitCode": "g+IlHSPp$M#opuAT\"1",
          "qualityControl": "Z"
        }
      }
    }
  ],
  "observationStations": [
    "string"
  ],
  "pagination": {
    "next": "string"
  }
}

{'@context': ['string'],
 'type': 'FeatureCollection',
 'features': [{'@context': ['string'],
   'id': 'string',
   'type': 'Feature',
   'geometry': {'type': 'Point', 'coordinates': [0, 0], 'bbox': [0, 0, 0, 0]},
   'properties': {'@context': ['string'],
    'geometry': 'string',
    '@id': 'string',
    '@type': 'wx:ObservationStation',
    'elevation': {'value': 0,
     'maxValue': 0,
     'minValue': 0,
     'unitCode': 'e0@A@9JrhssKVm:/+uD$8iW0v&+`jT*)Kdh<S@)R4`>Zf"X.|_1%=85_mxd-b@>Qo ep^fJe\\.uuEA9EhXXd',
     'qualityControl': 'Z'},
    'stationIdentifier': 'string',
    'name': 'string',
    'timeZone': 'string',
    'provider': 'string',
    'subProvider': 'string',
    'forecast': 'string',
    'county': 'string',
    'fireWeatherZone': 'string',
    'distance': {'value': 0,
     'maxValue': 0,
     'minValue': 0,
     'unitCode': 'uc:0k%4=i.gN%!{nFJfyl',
     'qualityControl': 'Z'},
    'bearing': {'value': 0,
     'maxValue': 0,
     'minValue': 0,
     'unitCode': 'g+IlHSP

In [94]:
if 'features' in data:
   print("Top Level Keys:", data.keys())

if 'features' in data:
    print(f"Number of observations found: {len(data['features'])}")

Top Level Keys: dict_keys(['@context', 'type', 'features', 'pagination'])
Number of observations found: 170


In [95]:
import pandas as pd

# Extract only 'properties' for each observation into a list
list_of_obs = [obs['properties'] for obs in observations]

# Create DataFrame
df = pd.DataFrame(list_of_obs)

# Since temperature and precipitation are dictionaries extract the 'value'
df['temp_c'] = df['temperature'].apply(lambda x: x['value'] if x else None)
df['precip_mm'] = df['precipitationLastHour'].apply(lambda x: x['value'] * 1000 if x and x['value'] is not None else 0)

# Keep main columns
df = df[['timestamp', 'temp_c', 'precip_mm']]

In [96]:
df.head()

timestamp  temp_c  precip_mm
0  2026-02-15T20:51:00+00:00     3.9        0.0
1  2026-02-15T19:51:00+00:00     3.3        0.0
2  2026-02-15T18:51:00+00:00     3.3        0.0
3  2026-02-15T17:51:00+00:00     3.3        0.0
4  2026-02-15T16:51:00+00:00     4.4        0.0

In [97]:
df.describe()

temp_c    precip_mm
count  170.000000   170.000000
mean    -0.448824     8.823529
std      4.562521   115.044748
min    -12.200000     0.000000
25%     -2.200000     0.000000
50%      0.600000     0.000000
75%      2.200000     0.000000
max      7.800000  1500.000000

In [86]:
# Setup NYC Central Park Station
station_id = "KNYC"
url = f"https://api.weather.gov/stations/{station_id}/observations"
headers = {'User-Agent': 'ClimateDataAssignment/1.0 (myemail@example.com)'}

# Fetch the data
response = requests.get(url, headers=headers)
data = response.json()

# NWS API 'features' is list of recent observations
observations = data['features']

In [84]:
# 1. Extract and print all dates and temperatures (Celsius, since this the is unit used in data set)

In [85]:
print("Date, Temperature (°C)")
for obs in observations:
    # Extract timestamp and temperature 
    timestamp = obs['properties']['timestamp']
    temp_c = obs['properties']['temperature']['value']
    
    # Formatting date (YYYY-MM-DD)
    date_only = timestamp[:10]
    
    if temp_c is not None:
        print(f"{date_only}, {temp_c}°C")

Date, Temperature (°C)
2026-02-15, 3.9°C
2026-02-15, 3.3°C
2026-02-15, 3.3°C
2026-02-15, 3.3°C
2026-02-15, 4.4°C
2026-02-15, 3.9°C
2026-02-15, 3.9°C
2026-02-15, 3.9°C
2026-02-15, 2.2°C
2026-02-15, 0.6°C
2026-02-15, 1.1°C
2026-02-15, 1.7°C
2026-02-15, 2.2°C
2026-02-15, 3.3°C
2026-02-15, 3.3°C
2026-02-15, 3.3°C
2026-02-15, 3.3°C
2026-02-15, 3.9°C
2026-02-15, 4.4°C
2026-02-15, 5°C
2026-02-15, 5°C
2026-02-14, 5.6°C
2026-02-14, 6.1°C
2026-02-14, 7.2°C
2026-02-14, 7.8°C
2026-02-14, 7.2°C
2026-02-14, 7.2°C
2026-02-14, 6.7°C
2026-02-14, 4.4°C
2026-02-14, 2.8°C
2026-02-14, 1.7°C
2026-02-14, 0.6°C
2026-02-14, 0°C
2026-02-14, 0.6°C
2026-02-14, 0°C
2026-02-14, -1.1°C
2026-02-14, -1.7°C
2026-02-14, -1.7°C
2026-02-14, -1.1°C
2026-02-14, -0.6°C
2026-02-14, -0.6°C
2026-02-14, -0.6°C
2026-02-14, -0.6°C
2026-02-14, -0.6°C
2026-02-14, 0°C
2026-02-14, 0.6°C
2026-02-14, 0.6°C
2026-02-14, 1.1°C
2026-02-13, 2.2°C
2026-02-13, 2.8°C
2026-02-13, 2.8°C
2026-02-13, 3.3°C
2026-02-13, 2.8°C
2026-02-13, 2.2°C
2026-0

In [87]:
# 2. Calculate average temperature (Celsius)

In [88]:
total_temp = 0
count = 0

for obs in observations:
    temp_c = obs['properties']['temperature']['value']
    if temp_c is not None:
        total_temp += temp_c
        count += 1

if count > 0:
    avg_temp = total_temp / count
else:
    avg_temp = 0

print(f"\nAverage temperature: {avg_temp:.2f}°C")


Average temperature: -0.45°C


In [46]:
# 3. Find days with precipitation

In [91]:
print("\nDays with precipitation:")
seen_dates = set()

for obs in observations:
    # precipitationLastHour is provided in meters by the API, will convert m to mm (*1000)
    precip_raw = obs['properties']['precipitationLastHour']['value']
    date_str = obs['properties']['timestamp'][:10]
    
    if precip_raw is not None and precip_raw > 0:
        if date_str not in seen_dates:
            precip_mm = precip_raw * 1000  # Converting meters to mm
            print(f"{date_str}: {precip_mm:.2f} mm")
            seen_dates.add(date_str)

print("\nDays with precipitation:")
for obs in data['observations']:
        # Check if precipitation is greater than 0
    if obs['precipitation'] > 0:
        print(f"Date: {obs['date']} | Precipitation: {obs['precipitation']} in")


Days with precipitation:
2026-02-11: 1500.00 mm


In [122]:
### PART 2 Downloading Files with Python

In [100]:
# Provided Code

In [115]:
import pooch
import os

# Set up Pooch to download a file
# This example downloads a small air quality dataset
file_path = pooch.retrieve(
    url="https://github.com/pandas-dev/pandas/raw/main/doc/data/air_quality_no2.csv",
    known_hash=None
)

print("File downloaded to:", file_path)
print("File exists:", os.path.exists(file_path))

File downloaded to: /home/awk2145/.cache/pooch/458dad453f6a48e510cd544bef1854e3-air_quality_no2.csv
File exists: True


In [105]:
# 1. Verify the file was downloaded (5 points)

In [116]:
# Check the file size
file_size = os.path.getsize(file_path)
print(f"File size: {file_size} bytes")

# YOUR CODE HERE: open the file and count how many lines it has
line_count = 0 
with open(file_path, "r") as f:
    for line in f:
        line_count += 1

print(f"Number of lines: {line_count}")

File size: 31984 bytes
Number of lines: 1036


In [112]:
# 2. Download another file

In [117]:
# Using a Sea Ice Extent dataset from NOAA/NSIDC as it's a reliable climate source
my_url = "https://noaadata.apps.nsidc.org/NOAA/G02135/north/daily/data/N_seaice_extent_daily_v4.0.csv"
my_file = pooch.retrieve(url=my_url, known_hash=None)

print(f"\nCustom download successful: {my_file}")
print(f"Custom file size: {os.path.getsize(my_file)} bytes")


Custom download successful: /home/awk2145/.cache/pooch/bb1127e5cb9f20dc390ca706afac8248-N_seaice_extent_daily_v4.0.csv
Custom file size: 1827712 bytes


In [113]:
# 3. Create a data inventory

In [118]:
print("\nData Inventory:")
print("1. meteorites.csv - NASA meteorite landings")
print("2. air_quality_no2.csv - Air quality NO2 measurements")
# Adding the new climate file
print(f"3. {os.path.basename(my_file)} - NOAA Daily Northern Hemisphere Sea Ice Extent")


Data Inventory:
1. meteorites.csv - NASA meteorite landings
2. air_quality_no2.csv - Air quality NO2 measurements
3. bb1127e5cb9f20dc390ca706afac8248-N_seaice_extent_daily_v4.0.csv - NOAA Daily Northern Hemisphere Sea Ice Extent


In [121]:
### PART 3 Understanding NetCDF Metadata

In [124]:
# Provided Code

In [120]:
import requests

# OPeNDAP provides metadata in different formats
# We'll get basic info about a climate dataset

base_url = "http://iridl.ldeo.columbia.edu/expert/SOURCES/.NOAA/.NCEP/.CPC/.UNIFIED_PRCP/.GAUGE_BASED/.GLOBAL/.v1p0/.Monthly/.RETRO/.rain/dods"

# Get DDS (Dataset Descriptor Structure) - describes the structure
dds_url = base_url + ".dds"
response = requests.get(dds_url)

print("Dataset Structure:")
print(response.text[:500])  # Print first 500 characters

Dataset Structure:
Dataset {
    Float32 T[T = 324];
    Float32 Y[Y = 360];
    Float32 X[X = 720];
    Grid {
     ARRAY:
        Float32 rain[T = 324][Y = 360][X = 720];
     MAPS:
        Float32 T[T = 324];
        Float32 Y[Y = 360];
        Float32 X[X = 720];
    } rain;
} rain;



In [138]:
ds

<xarray.Dataset> Size: 336MB
Dimensions:  (T: 324, Y: 360, X: 720)
Coordinates:
  * T        (T) float32 1kB 228.5 229.5 230.5 231.5 ... 548.5 549.5 550.5 551.5
  * Y        (Y) float32 1kB -89.75 -89.25 -88.75 -88.25 ... 88.75 89.25 89.75
  * X        (X) float32 3kB 0.25 0.75 1.25 1.75 ... 358.2 358.8 359.2 359.8
Data variables:
    rain     (T, Y, X) float32 336MB ...
Attributes:
    Conventions:  IRIDL

In [ ]:
# 1. Identify dimensions and variables (5 points)
# Look at the DDS output above and answer:
# - What are the dimension names? 
    ## They are T (time in months), and location X (latitude) and Y (longitude)
# - What is the main variable name?
    ## rain
# - Write your answers in a markdown cell
    ##

# 1. Identify dimensions and variables (5 points)
# Look at the DDS output above and answer:
# - What are the dimension names? 
    ## They are T (time in months), and location X (latitude) and Y (longitude)
# - What is the main variable name?
    ## rain

In [128]:
df.describe()

temp_c    precip_mm
count  170.000000   170.000000
mean    -0.448824     8.823529
std      4.562521   115.044748
min    -12.200000     0.000000
25%     -2.200000     0.000000
50%      0.600000     0.000000
75%      2.200000     0.000000
max      7.800000  1500.000000

In [129]:
df.head()

timestamp  temp_c  precip_mm
0  2026-02-15T20:51:00+00:00     3.9        0.0
1  2026-02-15T19:51:00+00:00     3.3        0.0
2  2026-02-15T18:51:00+00:00     3.3        0.0
3  2026-02-15T17:51:00+00:00     3.3        0.0
4  2026-02-15T16:51:00+00:00     4.4        0.0

In [140]:
# 2. Get data attributes (5 points)
# DAS (Dataset Attribute Structure) contains metadata
das_url = base_url + ".das"
das_response = requests.get(das_url)

print("Dataset Attributes (First 1000 chars):")
print(das_response.text[:1000])

Dataset Attributes (First 1000 chars):
Attributes {
    X {
        String standard_name "longitude";
        Float32 pointwidth 0.5;
        Int32 gridtype 1;
        String units "degree_east";
    }
    T {
        Float32 pointwidth 1.0;
        String calendar "360";
        Int32 gridtype 0;
        String units "months since 1960-01-01";
    }
    Y {
        String standard_name "latitude";
        Float32 pointwidth 0.5;
        Int32 gridtype 0;
        String units "degree_north";
    }
    rain {
        Int32 pointwidth 0;
        String standard_name "lwe_precipitation_rate";
        Float32 file_missing_value -999.0;
        String history "Boxes with less than 0.0% dropped";
        Float32 missing_value NaN;
        String units "mm/day";
        String long_name "Monthly Precipitation";
    }
NC_GLOBAL {
    String Conventions "IRIDL";
}
}



In [127]:
# 3. Document what you learned (5 points)
# In a markdown cell, write:
# - What does this dataset contain?
# - What time period does it cover?
# - What geographic region does it cover?
# - What are the units of the main variable?
# Find this info in the DAS output

In [143]:
import pandas as pd

# 1. Define the starting point (the 'epoch') from your DAS output
start_point = pd.to_datetime("1960-01-01")

# 2. These are the typical values found in the 'actual_range' of the RETRO dataset
# '228' represents Jan 1979 (228 months after Jan 1960)
# '551' represents Dec 2005 (551 months after Jan 1960)
data_start_val = 228
data_end_val = 551

# 3. Calculate the actual dates
actual_start_date = start_point + pd.DateOffset(months=data_start_val)
actual_end_date = start_point + pd.DateOffset(months=data_end_val)

print(f"The numerical value {data_start_val} converts to: {actual_start_date.strftime('%B %Y')}")
print(f"The numerical value {data_end_val} converts to: {actual_end_date.strftime('%B %Y')}")

The numerical value 228 converts to: January 1979
The numerical value 551 converts to: December 2005


# - What does this dataset contain?
monthly global precipiation amounts
# - What time period does it cover?
Jan 1979 - Dec 2005 OR months since 1960-01-01
# - What geographic region does it cover?
global areas from 89.75°N to 89.75°S in latitude, and 0.25°E to 359.75°E in longitude
# - What are the units of the main variable?
mm/day
